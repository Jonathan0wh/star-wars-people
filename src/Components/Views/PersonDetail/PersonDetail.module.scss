.column_container {
  padding: 16px;
  max-width: 900px;
}
