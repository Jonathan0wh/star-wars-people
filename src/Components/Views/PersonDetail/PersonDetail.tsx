import { useAppSelector } from "hooks/react-redux";
import { Col, Row, Space, Typography } from "antd";
import FilmCard from "Components/Shared/FilmCard";

import styles from "./PersonDetail.module.scss";

/**
 * Display a person's detail
 *
 * @returns A person's detail in a row, including all his film cards
 */
function PersonDetail() {
  const { Title, Text } = Typography;

  const { name, birthYear, gender, filmIDs } = useAppSelector(
    (state) => state.personDetail
  );

  return !!name ? (
    <Row className={styles.column_container}>
      <Col className={styles.column_container}>
        <Row>
          <Col>
            <Title level={4}>Person Detail</Title>
          </Col>
        </Row>

        <Row>
          <Col>
            <Text strong>Name:&nbsp;</Text>
          </Col>
          <Col>
            <Text>{name}</Text>
          </Col>
        </Row>

        <Row>
          <Col>
            <Text strong>Birth year:&nbsp;</Text>
          </Col>
          <Col>
            <Text>{birthYear}</Text>
          </Col>
        </Row>

        <Row>
          <Col>
            <Text strong>Gender:&nbsp;</Text>
          </Col>
          <Col>
            <Text>{gender}</Text>
          </Col>
        </Row>

        <Row>
          <Col>
            <Text strong>List of films:</Text>
          </Col>
        </Row>
        <Space wrap>
          <Row gutter={[16, 16]}>
            {filmIDs.map((filmID) => (
              <FilmCard key={filmID} id={filmID} />
            ))}
          </Row>
        </Space>
      </Col>
    </Row>
  ) : null;
}

export default PersonDetail;
