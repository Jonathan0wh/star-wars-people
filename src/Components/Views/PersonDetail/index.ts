export { default } from "./PersonDetail";

export { default as personDetailReducer, updatePersonDetail } from "./reducer";
