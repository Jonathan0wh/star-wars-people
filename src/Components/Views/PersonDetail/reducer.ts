import { createSlice, PayloadAction } from "@reduxjs/toolkit";

interface PersonDetailState {
  name: string;
  birthYear: string;
  gender: string;
  filmIDs: Array<string>;
}

const initialState: PersonDetailState = {
  name: "",
  birthYear: "",
  gender: "",
  filmIDs: [],
};

const personDetailSlice = createSlice({
  name: "personDetail",
  initialState,
  reducers: {
    updatePersonDetail: (state, action: PayloadAction<PersonDetailState>) => {
      return action.payload;
    },
  },
});

export const { updatePersonDetail } = personDetailSlice.actions;

export default personDetailSlice.reducer;
