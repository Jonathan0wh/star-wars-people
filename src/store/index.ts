export { store, persistor } from "./store";

export type { RootState, AppDispatch } from "./store";
