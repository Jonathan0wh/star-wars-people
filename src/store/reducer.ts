import { combineReducers } from "redux";
import { swApi } from "services/swApi";

export const rootReducer = combineReducers({
  [swApi.reducerPath]: swApi.reducer,
});
