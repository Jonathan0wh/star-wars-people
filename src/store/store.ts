import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/query";
import { personDetailReducer } from "Components/Views/PersonDetail";
import { swApi } from "services/swApi";

export const store = configureStore({
  reducer: {
    [swApi.reducerPath]: swApi.reducer,

    personDetail: personDetailReducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(swApi.middleware),
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
