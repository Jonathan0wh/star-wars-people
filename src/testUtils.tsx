// test-utils.js
import React from "react";
import { applyMiddleware, createStore } from "redux";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import { render as rtlRender, RenderOptions } from "@testing-library/react";
// Import your own reducer
import { rootReducer } from "store/reducer";

type renderProps = {
  initialState?: object;
  store?: ReturnType<typeof createStore>;
} & RenderOptions;

function render(
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  {
    initialState,
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  }: renderProps = {}
) {
  return rtlRender(ui, {
    wrapper: ({ children }: { children?: React.ReactNode }) => (
      <Provider store={store}>{children}</Provider>
    ),
    ...renderOptions,
  });
}

// re-export everything
export * from "@testing-library/react";
// override render method
export { render };
