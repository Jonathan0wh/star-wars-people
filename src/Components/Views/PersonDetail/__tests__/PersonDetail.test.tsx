import { store } from "store";
import { render } from "testUtils";

import PersonDetail from "../PersonDetail";

describe("View Component: PersonDetail", () => {
  const ref: React.Ref<HTMLDivElement> = {
    current: null,
  };

  it("should render by default", () => {
    const component = render(<PersonDetail sectionRef={ref} />);

    const tree = component.asFragment();
    expect(tree).toMatchSnapshot();
  });

  it("should render with correct store data", () => {
    const component = render(<PersonDetail sectionRef={ref} />, { store });

    const tree = component.asFragment();
    expect(tree).toMatchSnapshot();
  });
});
