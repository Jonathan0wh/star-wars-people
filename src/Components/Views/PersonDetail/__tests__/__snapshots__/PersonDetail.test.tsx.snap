// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Component: PersonDetail should render by default 1`] = `
<DocumentFragment>
  <div
    class="ant-row row_container"
  >
    <div
      class="ant-col"
    />
  </div>
</DocumentFragment>
`;

exports[`View Component: PersonDetail should render with correct store data 1`] = `
<DocumentFragment>
  <div
    class="ant-row row_container"
  >
    <div
      class="ant-col"
    />
  </div>
</DocumentFragment>
`;
